services/teamcity/teamcity-coverage.service.js

```diff
@@ -1,92 +1,111 @@
-        staticPreview: { message: '55%', color: 'yellow' },
+        title: 'TeamCity Coverage (CodeBetter)',
+        pattern: ':buildId',
+        namedParams: {
+          buildId: 'bt428',
+        },
+        staticPreview: this.render({
+          coverage: 82,
+        }),
+      },
+      {
+        title: 'TeamCity Coverage',
+        pattern: ':protocol/:hostAndPath/s/:buildId',
+        namedParams: {
+          protocol: 'https',
+          hostAndPath: 'https/teamcity.jetbrains.com',
+          buildId: 'bt428',
+        },
+        staticPreview: this.render({
+          coverage: 95,
+        }),
       },
     ]
   }
 
-  static get defaultBadgeData() {
-    return {
-      label: 'coverage',
-    }
+  async handle({ protocol, hostAndPath, buildId }) {
+    // JetBrains Docs: https://confluence.jetbrains.com/display/TCD18/REST+API#RESTAPI-Statistics
+    const buildLocator = `buildType:(id:${buildId})`
+    const apiPath = `app/rest/builds/${encodeURIComponent(
+      buildLocator
+    )}/statistics`
+    const data = await this.fetch({
+      protocol,
+      hostAndPath,
+      apiPath,
+      schema: buildStatisticsSchema,
+    })
+
+    const { coverage } = this.transform({ data })
+    return this.constructor.render({ coverage })
   }
 
-  static registerLegacyRouteHandler({ camp, cache }) {
-    camp.route(
-      /^\/teamcity\/coverage\/(.*)\.(svg|png|gif|jpg|json)$/,
-      cache((data, match, sendBadge, request) => {
-        const buildType = match[1] // eg, `bt428`.
-        const format = match[2]
-        const apiUrl = `http://teamcity.codebetter.com/app/rest/builds/buildType:(id:${buildType})/statistics?guest=1`
-        const badgeData = getBadgeData('coverage', data)
-        request(
-          apiUrl,
-          { headers: { Accept: 'application/json' } },
-          (err, res, buffer) => {
-            if (err != null) {
-              badgeData.text[1] = 'inaccessible'
-              sendBadge(format, badgeData)
-              return
-            }
-            try {
-              const data = JSON.parse(buffer)
-              let covered
-              let total
+  transform({ data }) {
+    let covered, total
 
-              data.property.forEach(property => {
-                if (property.name === 'CodeCoverageAbsSCovered') {
-                  covered = property.value
-                } else if (property.name === 'CodeCoverageAbsSTotal') {
-                  total = property.value
-                }
-              })
+    for (const p of data.property) {
+      if (p.name === 'CodeCoverageAbsSCovered') {
+        covered = +p.value
+      } else if (p.name === 'CodeCoverageAbsSTotal') {
+        total = +p.value
+      }
 
-              if (covered === undefined || total === undefined) {
-                badgeData.text[1] = 'malformed'
-                sendBadge(format, badgeData)
-                return
-              }
+      if (covered !== undefined && total !== undefined) {
+        break
```

Comment:
You could simplify the logic by moving the return statement here, and then unconditionally throw if the loop completes without reaching this.

Reply:
Yup, good idea